File: Easy/1.Two Sum/02_attempt.py
#================================
# Working process:
#   1. Iterate through nums with double loop variable
#   2. Set comp variable defined as difference of target and list nums's element
#   3. Return the list of comp and i if comp is in the dictionary memo
#   4. Else, put that element in the memo as a key
# Refinement: Shorter runtime by using only one loop
# TakeAway: Learning how to apply library data type
#================================


class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        memo = {}

        for i, num in enumerate(nums):
            comp = target - num

            if comp in memo:
                return[memo[comp],i]
            else:
                memo[num] = i

File: Easy/1.Two Sum/02_attempt.cpp
//================================
// Working process:
//   1. Iterate through nums
//   2. Iterate through nums once again from where we left off
//   3. Check if sum of those two elements of nums are equal to target
//   4. Return the indices if found
// Refinement: variable comp is declared inside the for loop to limit its scope and improve readability.
// TakeAway: Learning to write code in C++ and using unordered_map to store the pairs of numbers and their indices for faster lookup.
//================================


#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> pairs;

        for(int i{0}; i<nums.size(); ++i){
            int comp = target - nums[i];
            if(pairs.contains(comp)){
                return {pairs[comp],i};
            }
            else{
                pairs[nums[i]] = i;
            }
        }

        return {};
    }
};